File: codeforces/bin_serch/step2/rectangles.cpp
#include <bits/stdc++.h>

using namespace std;
#define ll long long

ll w, h, n;

int main() {
	scanf("%lld %lld %lld", &w, &h, &n);
	ll lower = 0;
	ll upper = 1e9;
	while (lower < upper) {
		ll mid = (lower + upper) / 2;
		if ((mid/h) * (mid/w) < n)
			lower = mid+1;
		else
			upper = mid;
	}
	printf("%lld\n", lower);
}

File: codeforces/bin_serch/step2/rectangles.py
w, h, n = map(int, input().split())
lower = 0
upper = int(1e14)
while (lower < upper):
	mid = (lower + upper) // 2
	if (mid//w) * (mid//h) < n:
		lower = mid+1
	else:
		upper = mid
print(lower)
